src/main/java/com/xpn/xwiki/plugin/lucene/IndexRebuilder.java

```diff
@@ -122,273 +146,226 @@
-      // context.setLanguage(this.context.getLanguage());
-      // context.setDatabase(this.context.getDatabase());
-      // context.put("org.xwiki.component.manager.ComponentManager", this.context
-      // .get("org.xwiki.component.manager.ComponentManager"));
-      context = getContext().getXWikiContext();
-      // For example, we definitely don't want to use the same hibernate session...
-      context.remove("hibsession");
-      context.remove("hibtransaction");
-      // This is also causing serious problems, as the same xcontext gets shared between
-      // threads and causes the hibernate session to be shared in the end. The vcontext is
-      // automatically recreated by the velocity renderer, if it isn't found in the
-      // xcontext.
-      context.remove("vcontext");
-
-      // The original request and response should not be used outside the actual request
-      // processing thread, as they will be cleaned later by the container.
-      context.setRequest(null);
-      context.setResponse(null);
-
-      rebuildIndex();
-    } catch (Exception e) {
-      LOGGER.error("Error in lucene rebuild thread: {}", e.getMessage(), e);
-    } finally {
-      if (context != null) {
-        context.getWiki().getStore().cleanUp(context);
+  public synchronized IndexRebuildFuture startIndexRebuild(final EntityReference filterRef) {
+    rebuildQueue.removeIf(CompletableFuture::isDone);
+    return getQueuedRebuild(filterRef).orElseGet(() -> {
+      IndexRebuildFuture newFuture = new IndexRebuildFuture(filterRef);
+      rebuildQueue.add(newFuture);
+      try {
+        rebuildIndexAsync(newFuture);
+      } catch (Exception exc) {
+        newFuture.completeExceptionally(exc);
       }
-      rebuildInProgress.set(false);
-    }
+      return newFuture;
+    });
+  }
 
-    LOGGER.info("Lucene index rebuild done");
+  protected void rebuildIndexAsync(final IndexRebuildFuture future) {
+    final EntityReference filterRef = future.getReference();
+    final Directory directory = expectIndexUpdater().getDirectory();
+    WikiReference wikiRef = References.extractRef(filterRef, WikiReference.class).get();
+    ContextExecutor.executeInWiki(wikiRef, () -> CompletableFuture.runAsync(
+        new AbstractXWikiRunnable(XWikiContext.EXECUTIONCONTEXT_KEY, createJobContext()) {
+
+          @Override
+          protected void runInternal() {
+            LOGGER.info("[{}] - started", logRef(filterRef));
+            try (IndexSearcher searcher = new IndexSearcher(directory, true)) {
+              long count = rebuildIndex(searcher, filterRef);
+              LOGGER.info("[{}] - finished: {}", logRef(filterRef), count);
+              future.complete(count);
+            } catch (InterruptedException exc) {
+              future.completeExceptionally(exc);
+              Thread.currentThread().interrupt();
+            } catch (Exception exc) {
+              LOGGER.error("[{}] - failed: {}", filterRef, exc.getMessage(), exc);
```

Comment:
Why do we need to print the Message additionally to the complete exception?

Reply:
The future is returned to the caller `startIndexRebuild` or may be polled while in the `rebuildQueue` to check the status of the requested rebuild. There is no guarantee that the exception will be logged anywhere after `completeExceptionally`.

Reply:
I've however added missing error logs before the other `completeExceptionally` calls.

Reply:
it's not needed for error, since we'll always print the stacktrace on this level. useful for warn level or below. hence I removed the message logging.